### ISSUE TITLE
apnProvider.send() never hangs and never comes back

### ISSUE THREAD

**Issue by trx222** [node-apn/node-apn#665]
For fast delivery of our notifications I start multiple send Processes and splitt up the receiver tokens to them. In my project actually we using 8 processes. From time to time on of the processes stops working in apnProvider.send() and never comes back. I give an array with up to 100 receivers but then I don't know what happens - then() is never called.

How can I find out, whats the problem? Are there any debug possibilities? I tested with different node version with node-apn 2.2

Sometimes I also get 
"MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 wakeup listeners added to [Connection]. Use emitter.setMaxListeners() to increase limit" 

but the it again works for hours and days ... later it stops every few notifications. 

thanks in advise

**Comment by trx222** [comment]
I added now some debug output to client.js and found out that in 
Client.prototype.write 
I get in this failure cases less results then requests sent. I send a notification to 18 devices and get back 17 times a 200 and then nothing happens, no error, no timeout ... it just waits infinitely.

Is there a way to catch this, with a timeout and try a resend in that case?

But better question - how and why can this happen :-(

**Comment by d0972058277** [comment]
My Typescript Code.

```
class _TimeoutPromise {
    Set<T>(promise: Promise<T>, ms: number, err_message?: string): Promise<T> {
        // Create a promise that rejects in <ms> milliseconds
        let timeout = new Promise<T>((resolve, reject) => {
            let id = setTimeout(() => {
                clearTimeout(id);
                if (err_message === null || err_message === undefined) reject('Timed out in ' + ms + 'ms.');
                reject(`Timed out in ${ms} ms.\n${err_message}`);
            }, ms);
        });

        // Returns a race between our timeout and the passed in promise
        return Promise.race([promise, timeout]);
    }
}
export const TimeoutPromise = new _TimeoutPromise();
```
```
import Config from 'config';
import { TimeoutPromise } from 'TimeoutPromise'

// Write Your Provider.send() Promise.
let options : ProviderOptions = { token: Config.get(`PushSetting.APN.Token`) };
let provider = new Provider(options);
let sendPromise = provider.send(notification,tokens);

let result = await TimeoutPromise.Set(sendPromise, Config.get('PushSetting.ApnTimeout') as number)
	.catch(async err => {
		// My business use, just like "console.log".
		// Notify APN is timeout to me.
        let message = `Node-Apn超過${Config.get('PushSetting.ApnTimeout')}ms沒有回應。`;
		await LineNotify.Info(message);

        // Write Your Recursive Code.
    });
```

**Comment by trx222** [comment]
Hi. Thank you for your code. I also have added a simple timeout, works, but I really want to know and avoid the real problem. If I have a bulk array of 100 devices and it stops somewhere, I don't know which device has received it and which needs a resend. I also tried now 2.1.5 again but it also happens with this. Strange because the job worked now over years :-(

**Comment by d0972058277** [comment]
Sorry, my English is not good.
Let me use google translate to talk with you.

I use the [#664](https://github.com/node-apn/node-apn/issues/664) module.
https://github.com/parse-community/node-apn
https://www.npmjs.com/package/@parse/node-apn
Nodejs Version: 10.16.3
Nodejs (10.16.3) has the module of http2.

"MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 wakeup listeners added to [Connection]. Use emitter.setMaxListeners() to increase limit"
This error message will happen when I send more than 10 tokens at a time.
I guess this is caused by this [code](https://github.com/node-apn/node-apn/blob/38a357ed0c153aad09c2857e48a710527e685bfc/lib/provider.js#L29).
In order to solve this problem, you need to adjust the MaxListeners to the maximum number of tokens you will send at one time.
So if I want to send 1000 tokens, I need to adjust MaxListeners to 1000.
Although this is not the best way.
```
require('events').EventEmitter.defaultMaxListeners = 1000;
```
If there are more than 1000? A batch of 1000 tokens is required.

I don't know how the "events" timeout will happen.
If you want to improve this, you need a Pull Request to make adjustments to this [code](https://github.com/node-apn/node-apn/blob/38a357ed0c153aad09c2857e48a710527e685bfc/lib/provider.js#L29).
Add an Idle array to store those overtime pushes, and you'll know which tokens need to be resent.

**Comment by trx222** [comment]
I am on testing the parse-community version and will post the results here. Thanks for this advise.

Regarding the EventEmitter Leak - in this case I would expect that this error comes always with so many devices, but thats not the case ... it also just comes some times, some days with a lot of notification I don't see it ... other days it comes multiple times.

**Comment by eladnava** [comment]
Just reproduced this issue in `@parse/node-apn` fork as well.

**Comment by trx222** [comment]
Yes but it seems that the "hang-problem" is solved in there code. I just got back for a bulk of devices "VError: stream ended unexpectedly" what I never had seen before with node-apn version here. I am able to catch this and send again the message to the devices.

**Comment by eladnava** [comment]
I'm going to stick with `@parse/node-apn` as well as it seems better maintained for now.

**Comment by TysonAndre** [comment]
I wonder if https://nodejs.org/api/http2.html#http2_server_settimeout_msecs_callback would help - After this was refactored to use the native http2 module, the socket timeouts and connection timeouts were no longer set by default, and there's no documentation in docs/ of how they should be used (as far as I could tell)

I'm concerned that requests to send() could hang forever (until the server ran out of memory?). Calling https://nodejs.org/api/http2.html#http2_server_settimeout_msecs_callback may help with that (after connecting to the server) - the default request timeout seems to be used by node-apn, and that default is 0 (no timeout).

Then again, the issue could be something else
